Line search

RikVoorhaar opened the pull request:
I made a Riemannian line search optimizer with strong Wolfe conditions. 

It's not yet perfect. I think it makes some redundant calls to `closure` during stepping, and when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,249 @@
+            params, defaults, stabilize=stabilize
+        )
+        self._params = self.param_groups[0]["params"]
+        self.c1 = self.param_groups[0]["c1"]
```
why is it global? Is it possible to have different constants between param groups?

RikVoorhaar replied:
`stabilize` should probably be per parameter group, but I don't see a way to implement different line search constants for different parameter groups.

ferrine replied:
> `stabilize` should probably be per parameter group

Fair point, I missed that

ferrine replied:
> I don't see a way to implement different line search

That means to avoid misunderstanding between user and developer we need some checks for that or do not set is per parameter group in the first place

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,249 @@
+                continue
+
+            # For some reason using the metric for inner product gives wrong results,
+            # therefore us euclidean product instead.
```
what was wrong there?

RikVoorhaar replied:
I now realized what went wrong probably. If using `manifold.inner(point, -state["new_grad"], state["grad_retr"])` at this point in the code, then the point is wrong. We need to compute this quantity before we roll back point at the end of the function `_phi`, and doing this seems to work well. I will modify the code.

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,249 @@
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norms.append(grad.norm().item())
```
do you really need Euclidean norm here and not Riemannian one?

RikVoorhaar replied:
No, Riemannian is better, thank you for the observation. I will update this.

ferrine replied:
btw, all norm computations are broadcastable and can factorize across product manifold. if you have a product manifold, you may better have more than one grad norm. 

```
# pseudocode
point = Sphere(10, 3)  # 10 3d spheres
inner = point.inner(v1, v2)
inner.shape
(10,)
```

Does having a factorized inner product makes sense in this optimizer? In Adam this affected adaptive term in a good way

RikVoorhaar replied:
I just need the squared sum of gradient norms to perform the line search, so this isn't necessary. But this is good to keep in mind if implementing geometric conjugate gradient (which seems useful).

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -186,12 +190,11 @@ def init_loss(self):
             # project gradient onto tangent space
             grad = manifold.egrad2rgrad(point, grad)
-            grad_norms.append(grad.norm().item())
+            grad_norm = manifold.norm(point, grad)
```
Be careful, norm and inner reduce only manifold dimensions, not batch ones. Was it what you meant?

RikVoorhaar replied:
Good point. I did not design this optimizer with batches in mind (e.g. line search doesn't really make sense in the stochastic gradient descent setting). I will alter the optimizer so that it works correctly with batches of tensors.

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -130,6 +138,8 @@ def phi_(self, step_size):
 
             state["new_grad"] = grad
 
+            state["der_phi"] = manifold.inner(point, -grad, state["grad_retr"])
```
Same as comment below about reduction dimensions, you better use sum after calling norm

ferrine commented:
I meant batches of tensors that are product manifolds

On Thu, 16 Jul 2020, 13:28 Rik Voorhaar, <notifications@github.com> wrote:

> *@RikVoorhaar* commented on this pull request.
> ------------------------------
>
> In geoopt/optim/rwolfe.py
> <https://github.com/geoopt/geoopt/pull/140#discussion_r455686408>:
>
> > @@ -186,12 +190,11 @@ def init_loss(self):
>              state = self.state[point]
>              # project gradient onto tangent space
>              grad = manifold.egrad2rgrad(point, grad)
> -            grad_norms.append(grad.norm().item())
> +            grad_norm = manifold.norm(point, grad)
>
> Good point. I did not design this optimizer with batches in mind (e.g.
> line search doesn't really make sense in the stochastic gradient descent
> setting). I will alter the optimizer so that it works correctly with
> batches of tensors.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/geoopt/geoopt/pull/140#discussion_r455686408>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZJX3X5RFB3XQMPUCUXIN3R33I45ANCNFSM4OXMZUXQ>
> .
>

RikVoorhaar commented:
> I meant batches of tensors that are product manifolds

In that case I don't think I understand what you mean. Without modification it seems to work just fine with tensors belonging to a `ProductManifold`. Either way, I modified the code to use `sum` in the cases you mentioned, and it now seems to work just fine with batches of tensors as well.

RikVoorhaar commented:
I figured that adding line search based on Armijo backtracking is basically free, so I added this functionality. 
Might also add weak wolfe conditions. Adding support for weak Wolfe conditions, and Hager-Zhang line search would also be nice, but requires more effort. Same goes for implementing conjugate gradient methods.

ferrine commented:
You mentioned 
> when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps

Is this problem investigated?

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,280 @@
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norm = torch.sum(manifold.norm(point, grad)).item()
```
Norm here is computed incorrect, I guess you meant
```suggestion
            grad_norm = torch.sum(manifold.inner(point, grad)).pow(.5).item()
```

RikVoorhaar replied:
> Norm here is computed incorrect, I guess you meant

what is the difference between the two? Shouldn't manifold.norm do exactly this?

ferrine commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,280 @@
+
+            state["new_grad"] = grad
+
+            state["der_phi"] = torch.sum(
```
btw, what happens in case you have more than one parameter? In here you compute norms, etc, per parameter. But from the conversation, I felt like you need a global norm, global inner product

RikVoorhaar replied:
The `state["new_grad"]` isn't used, but should be. If the step we take is the one suggested by the line search, then `state["new_grad"]` already contains the gradient of the loss function with respect to this parameter, and we can omit a call to closure(). I had this idea in mind, but forgot to implement it. This does assume that the user does not alter the parameters in any way between consecutive calls of `.step()`, and I'm not whether such an assumption is necessarily good. 

`state["der_phi"]` is used in the function `derphi_`, where we just sum over all of the paremeters' `state["der_phi"]`. This sum is the global inner product we need. So, right, in principle we do not need to store it like this and we can just compute and store the sum from the get-go, but I figured storing this might not be bad since it tells us the contribution of a particular parameter to the gradient of the line search objective, which could be a useful metric for the user. (though this should be documented).

ferrine replied:
> This does assume that the user does not alter the parameters in any way between consecutive calls 

I believe this covers 99.9% usecases. This is also the assumption in all geoopt optimizers that have momentum support.

RikVoorhaar commented:
> You mentioned
> 
> > when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps
> 
> Is this problem investigated?

Yes, if this line search fails to find a new step it falls back to a step of a fixed size, which tends to be very large step. 
This should only happen in case the point is near a local minimum (or saddle point). Typically this might be the point where we halt the optimization process (or before this point already). Perhaps a good option would be to default to not doing a step at all, instead of using a large fallback step size, and simply warn the user that the convergence has been reached and additional steps are prone to numerical error.

ferrine commented:
I

> > You mentioned
> > > when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps
> > 
> > 
> > Is this problem investigated?
> 
> Yes, if this line search fails to find a new step it falls back to a step of a fixed size, which tends to be very large step.
> This should only happen in case the point is near a local minimum (or saddle point). Typically this might be the point where we halt the optimization process (or before this point already). Perhaps a good option would be to default to not doing a step at all, instead of using a large fallback step size, and simply warn the user that the convergence has been reached and additional steps are prone to numerical error.

Hmm, makes sense. Do you know any practical way to detect convergence of line search to implement here? Halting by default sounds good

RikVoorhaar commented:
> Hmm, makes sense. Do you know any practical way to detect convergence of line search to implement here? Halting by default sounds good
Convergence happens if the gradient of the line search objective is small (since this is nothing but the norm-squared of the function gradient). Unfortunately "small" is not a well-defined concept, so checking whether or not the optimization has converged is up to the user. From my experiments, the line search fails to find a new point mainly when it is close to a local minimum. 
I left it now that by default the optimizer does not take a step if the it can't find a good stepsize, but the user can call the `step` method with `force_step=True` to override this behavior.

ferrine commented:
There was a strange failure in travis, restarting build

ferrine commented:
oh it's doctest failing

RikVoorhaar commented:
> oh it's doctest failing

yes, I noticed this, but the error seemed too trivial to fix with a new commit.

RikVoorhaar commented:
travis errors are in `test_gyrovector_math.py` and `test_lorentz_math.py`, was getting the same thing locally.

ferrine commented:
Looks like pytest update

RikVoorhaar commented:
I'm now thinking of partially rewriting this class for more general behavior. 

Right now search direction of line search is fixed to be gradient direction. Instead we can use conjugate gradient search direction:

![CodeCogsEqn](https://user-images.githubusercontent.com/12045781/89323514-1b49bb00-d686-11ea-8abf-4bb8ebc0bf4e.gif)

[(Zhu-Sato suggests to use inverse retraction here instead of vector transport)](http://www.optimization-online.org/DB_FILE/2020/05/7798.pdf)

Steepest descent then boils down to setting beta = 0. Most popular choices of beta_{k+1} involve the same parameters:

![CodeCogsEqn (1)](https://user-images.githubusercontent.com/12045781/89323700-6ebc0900-d686-11ea-8bf2-7c9ec1eeba18.gif)

These parameters are all typically computed during the line search anyway, so it's free to use this. The idea is then to make all functions determining beta_{k+1} use the same API, so that it's easy for the user to write new functions. 

There are also many different ways of doing the line search itself, and there are many stopping criteria for the linesearch possible. Even in some particular cases exact line search is feasible. Right now it's not simple for the user to change the linesearch method to something new, so the class should be rewritten to make linesearch methods use the same API.

All in all, I think these modifications will make the optimizer more useful and general.

ferrine commented:
I'll fix pytest soon in master

ferrine commented:
You can rebase to fix pytest errors

RikVoorhaar commented:
One of tests is failing only on MacOS. This test is a bit prone to numerical errors, so I need to implement a better test. I have a list of test problems, most of which are less prone to numerical errors.

RikVoorhaar commented on geoopt/optim/rwolfe.py:
```diff
@@ -0,0 +1,689 @@
+
+            # Use retract to perform the step, and transport the search direction
+            if self.transport_search_direction:
+                search_transported = manifold.transp_follow_retr(
```
Using `transp_follow_retr` here seems less efficient in general than using `transp`.  However, for CanonicalStiefel for example, `transp` is not defined but `transp_follow_retr` is.

RikVoorhaar commented:
I added conjugate gradients and made it easier for the user to supply their own linesearch functions. I also fixed a number of small bugs, and found a couple points where I could make it a bit more efficient.

RikVoorhaar commented:
Fletcher Reeves scale parameter for conjugate gradient method seems to not work well with the orthogonal procrustes problem I use for testing. I need to investigate more carefully if there is no error in the implementation.

RikVoorhaar commented:
It seems the problem is solved by taking square norm as loss in the test, instead of just norm. In fact before the gradient norms didn't even go to zero.

ferrine commented:
Hi! Is it time for review or you have any ideas to improve docs?

RikVoorhaar commented:
I made some minor changes, most importantly I renamed rwolfe.py to rlinesearch.py, since this name is more appropriate (in fact it seems that in most practical cases armijo outperforms wolfe line search in the Riemannian setting). 

It should be ready for review. I don't know how to improve the docs any further.

ferrine commented:
I'll have look this week! Thanks!

On Mon, 7 Sep 2020, 14:04 Rik Voorhaar, <notifications@github.com> wrote:

> I made some minor changes, most importantly I renamed rwolfe.py to
> rlinesearch.py, since this name is more appropriate (in fact it seems that
> in most practical cases armijo outperforms wolfe line search in the
> Riemannian setting).
>
> It should be ready for review. I don't know how to improve the docs any
> further.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/geoopt/geoopt/pull/140#issuecomment-688251752>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZJX3T4JY7GIDWRTGANI7LSES43TANCNFSM4OXMZUXQ>
> .
>

ferrine reviewed (commented):
Looks good, you just need to merge master in this branch. And we are ready

RikVoorhaar commented:
done

ferrine commented:
Could you please fix the rest ?  There seem to be some unused imports in tests

RikVoorhaar commented:
should be good now

ferrine commented:
Great, thanks, that is solid
